BlueprintUI/Sources/Layout/GridRow.swift

```diff
@@ -0,0 +1,244 @@
+        ///   - isExactConstraint: Whether or not the constraint describes the row's size exactly. During measurement,
+        ///   it does not. During layout, it does.
+        ///   - items: Children to layout.
+        private func _frames(
+            in constraint: SizeConstraint,
+            isExactConstraint: Bool = false,
+            items: [(traits: Width, content: Measurable)]
+        ) -> [CGRect] {
+            var sizes: [CGSize] = (0..<items.count).map { _ in .zero }
+
+            // Group children by their sizing. Maintain child order by also storing index.
+            var absolutelySized: [(index: Int, width: CGFloat, content: Measurable)] = []
+            var proportionallySized: [(index: Int, proportion: CGFloat, content: Measurable)] = []
+            items.enumerated().forEach { (index, item) in
+                switch item.traits {
+                case .absolute(let width):
+                    absolutelySized.append((index, width, item.content))
+                case .proportional(let proportion):
+                    proportionallySized.append((index, proportion, item.content))
+                }
+            }
+
+            // Measure absolutely-sized children and record the total width they require.
+            var absoluteItemWidth: CGFloat = 0
+            absolutelySized.forEach { (index, width, content) in
+                absoluteItemWidth += width
+                let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                sizes[index] = CGSize(width: width, height: content.measure(in: constraint).height)
+            }
+
+            // Measure proportionally-sized children, allotting them a portion of the available space.
+            let availableWidth = constraint.width.maximum - (spacing * CGFloat(items.count - 1)) - absoluteItemWidth
+            // Determine the number of portions and the width of each portion.
+            let portionCount = proportionallySized.lazy.map { $0.proportion }.reduce(0, +)
+            let portion = availableWidth / portionCount
+            // Assign a proportional width to each proportionally-sized child.
+            proportionallySized.forEach { (index, proportion, content) in
+                let width = portion * proportion
+                let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                sizes[index] = CGSize(width: width, height: content.measure(in: constraint).height)
+            }
+
+            // Begin computing frames by reading the vertical alignment. Vertical alignment determines the y-origin
+            // and, in the `.fill` case, the height of the children.
+            //
+            // The tallest child will be used to size the row unless it is a layout pass, in which the row's height
+            // is already known.
+            let rowHeight = isExactConstraint
+                ? constraint.height.maximum
```

Comment:
Again, this is not going to behave well in unconstrained measurement, and probably won't play nicely if it's nested in other containers either (like if it's in a stack with other content).

In general I would recommend avoiding the `maximum` property except for some special uses, like comparing to a range.

Reply:
I think this should be safe to do in the `.exactly` case, which is what I'm getting at here.